### DISCUSSION TITLE
[RFC] Состояние скелета для каждого компонента

### DISCUSSION THREAD

**Discussion by lihachyov** [VKCOM/VKUI#2478] [Ideas] [answered]
Есть задача поддержать скелеты компонентов, отображающие серые прямоугольники как плейсхолдеры во время загрузки данных, предварительно показывая структуру элементов. 
 
Предлагается подумать насчёт реализации скелета. Возможно ли добавить state для каждого UI-компонента, когда данные в них могли бы отображаться схематично как на скриншоте?

<img width="1920" alt="EdkpK83WkAI1DWQ" src="https://user-images.githubusercontent.com/3250391/88265879-e9744400-ccd6-11ea-9685-1a00429c6b68.png">

**Comment by pfrankov** [comment]
@Lihachyov переоткрой, когда будет дизайн

**Comment by eolme** [comment]
@pfrankov так issue не закрываются, фича актуальна и не реализована

**Comment by Zaycevq** [comment]
Будем брать в работу командой Дизайна. Частично Скелетоны уже реализовывали, нужно будет эту логику перенести на компоненты VKUI
<img width="906" alt="Снимок экрана 2022-05-06 в 12 30 25" src="https://user-images.githubusercontent.com/91548592/167106011-07db77a5-298c-4693-90db-a0c8f48b4f58.png">

**Comment by inomdzhon** [comment]
Спасибо, @Zaycevq 🙏 

---

Надо будет взвесить насколько это усложнит код, если говорить о том, что для всех компонентов надо будет завести параметр меняющий состояния на скелетон.

Вижу пару вариантов для упрощения:
1. Если не ошибаюсь, скелетон зачастую отображает лишь грубый шаблон лэйаута, поэтому возможно стоит создать компонент `Skeleton`, который будет принимать параметры, помогающие визуально набросать тот или иной компонент. Да, прям как на скриншоте выше тогда не будет.
2. Делаем тот же `Skeleton`, но пробрасываем в компонент
    ```tsx
    <Component before={loading ? <Skeleton width={24} height={24} /> : <Avatar src="example.png" />}>
       {loading ? <Skeleton height={16} /> : data}
    </Component>
    ```

В общем, надо обсуждать)

**Comment by eugpoloz** [comment]
Как насчет превратить этот issue в дискуссию? Уже по технической реализации, так скажем. (% Почему: задача все равно слишком объемная и мы 100% будем ее дробить. Можем завести milestone "Скелетоны" и под него уже их подбивать.

@inomdzhon 
> Надо будет взвесить насколько это усложнит код, если говорить о том, что для всех компонентов надо будет завести параметр меняющий состояния на скелетон.

(Это еще может быть группа компонентов типа `AvatarSkeleton`, `TypographySkeleton` и так далее, если речь про несколько грубых атомарных компонентов.)

**Comment by SevereCloud** [comment, marked as answer]
Компонент [скелетона](https://vkcom.github.io/VKUI/#/Skeleton) появился в 6.1.0
